### PR TITLE
Consolidate ApiEnv and ProvisioningKey as 1 model

### PR DESCRIPTION
**Pull request by EdwardAndress**
ApiEnv was an array, dressed up as a model, and it was not possible to create new envs within the application.
The ability to create new environments will soon be required.
We considered convertin ApiEnv to a full model.
This would have meant a 1:1 relationship between ApiEnv and ProvisioningKey.
The only attribute on ApiEnv would have been #name
We felt, therefore, that it would make more sense to consolidate ApiEnv and ProvisioningKey into one model.

- Renamed ProvisioningKey to Environment
- Token and AccessRequest now belong to Environment
- Environment has many Tokens and AccessRequests
- Fixed broken tests and tested migration
- New cuke is commented out due to the way in which Cucumber is being used at present, which means cucumber scenarios write to the dev DB and database cleaner does not clean.

Still to be done:
- Remove or fix new cukes (current cucumber config doesn't allow conventional feature testing)
- Add the ability to change an environment's provisioning key (via `EnvironmentsController#update`)

### CODE REVIEW

**Review by himal-mandalia — COMMENTED** [review]
Looks good @EdwardAndress. We should also remove the unneeded `api_env` fields form `AccessRequest` and `Token` once the migration has been run. See inline comments.

File: db/migrate/20170810164131_rename_provisioning_keys_to_environments.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      env = Environment.find_by(name: token.api_env)
+      token.environment_id = env.id
+      token.save!
+    end
+
+    AccessRequest.all.each do |access_request|
+      env = Environment.find_by(name: access_request.api_env)
+      access_request.environment_id = env.id
+      access_request.save!
+    end
```
Comment (himal-mandalia) [inline db/migrate/20170810164131_rename_provisioning_keys_to_environments.rb:20]: This should probably be written as a:
```
def up
  ...
end

def down
  ...
end
```
In order to make it reversible. So the `up` method would apply the changes that migrate the data and remove the `api_env` from `AccessRequest` and `Token` and the `down` method would do the reverse of that - restore the `api_env` field and move the string values back into it. `change` method is good for simple migrations where Rails can figure out what the reverse/rollback should be.

File: features/managing_environments.feature [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Feature: Managing environments
```
Comment (himal-mandalia) [inline features/managing_environments.feature]: Lets discuss this - we can probably skip Cucumber for this altogether.

File: db/migrate/20170810164131_rename_provisioning_keys_to_environments.rb (hunk 1 above)
Comment (EdwardAndress) [reply db/migrate/20170810164131_rename_provisioning_keys_to_environments.rb:20]: Ah yes, I had intended to do this after testing the migration but forgot.  Thanks for pointing it out.

File: features/managing_environments.feature (hunk 2 above)
Comment (EdwardAndress) [reply features/managing_environments.feature]: Sure.  I'm OK either way.

File: db/migrate/20170810164131_rename_provisioning_keys_to_environments.rb (hunk 1 above)
Comment (EdwardAndress) [reply db/migrate/20170810164131_rename_provisioning_keys_to_environments.rb:20]: And, yes, I forgot to drop the api_env column on tokens and access_requests.  Well spotted.

File: app/models/environment.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class Environment < ApplicationRecord
+
+  has_many :tokens
```
Comment (himal-mandalia) [inline app/models/environment.rb]: `dependent: :nullify` to set the `environment_id` on dependent records to nil. Will have to ensure that token and access request template (index, show etc) handle displaying of nilled values.

Comment (himal-mandalia) [inline app/models/environment.rb]: `before_destroy` callback to revoke all tokens for an environment that is being removed.
```
  private
  
  def revoke_all_tokens!
    self.tokens.all.each &:revoke!
  end
```

Corresponding model spec to test all tokens for a destroyed environment are revoked:

`expect(token).to be_revoked`

**Comment by himal-mandalia** [comment]
@EdwardAndress one other point to consider: When the provisioning key on an environment is changed/updated it will invalidate existing tokens.

**Comment by EdwardAndress** [comment]
Thanks for the comments @himal-mandalia .  I think I have addressed them all.  One thing I'm not sure about is what to do with access_requests that are made against deleted environments.  As things stand, as an admin user, if you try to approve one you'll have to select one of the existing envs from a dropdown list on the next page.... this would mean we can grant access to an alternative environment, in those cases, which is maybe a good thing...  what do you think?

**Comment by EdwardAndress** [comment]
Also, I've amended my changes into the previous commit.  Is that what you would prefer?

**Comment by himal-mandalia** [comment]
@EdwardAndress looks good!

I think for now allowing outstanding access requests that are for an environment which is deleted can be fine - with that choice to then choose another environment on the form. This could be useful for when we switch a dev environment by deleting one and creating another.

Probably better to have them as a separate commit to allow the tracking of what's changed based on feedback etc. But it's fine.

Will test in a bit.

**Comment by EdwardAndress** [comment]
Conflict resolved @himal-mandalia

**Pull request merged**